### PR TITLE
CameraXBasic - simple androidTest for Preview and Surface usage

### PR DESCRIPTION
**Pull request by luncliff**
### Changes

It was a bit hard for me to try CameraX with the sample code. The PR adds a new test class which is shorter than the CameraFragment.kt.
 
### Coverage?

Tested with...

* Sony G8141, OS 9.0

### CODE REVIEW

**Comment by owahltinez** [comment]
Hi @luncliff thanks for this PR! We don't have any other code in Java for this sample, do you mind converting it to Kotlin? That way we can have a consistent codebase.

**Comment by luncliff** [comment]
@owahltinez thanks for the comment! 
Since I don't know Kotlin, so reformed the code with Android Studio. 

Any kind of review(especially for _readability_) will be helpful :)

File: CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    @Before
+    fun markCreated() {
+        registry = LifecycleRegistry(this)
+        registry!!.markState(Lifecycle.State.INITIALIZED)
+        registry!!.markState(Lifecycle.State.CREATED)
+    }
+
+    @After
+    fun markDestroyed() {
+        registry!!.markState(Lifecycle.State.DESTROYED)
```
Comment (owahltinez) [inline CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt:95]: Replace with `registry?.markState(...)`

File: CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    fun teardown() {
+        if (provider != null) provider!!.unbindAll()
+        if (reader != null) reader!!.close()
+        if (thread != null) thread!!.quit()
+        if (executor != null) executor!!.shutdown()
+    }
+
+    override fun getLifecycle(): Lifecycle {
+        return registry!!
+    }
```
Comment (owahltinez) [inline CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt:84]: Can be simplified to:

```
override fun getLifecycle() = registry
```

**Review by owahltinez — CHANGES_REQUESTED** [review]
Thanks for converting this to Kotlin! I have a few suggestions to make this code more idiomatic

File: CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    private var registry: LifecycleRegistry? = null
+    private var context: Context? = null
+    private var thread: HandlerThread? = null
+    private var executor: ExecutorService? = null
+    private var reader: ImageReader? = null
+    private var camera: Camera? = null
+    private var provider // requires main thread
+            : ProcessCameraProvider? = null
+    private var count = AtomicInteger(0)
```
[the remark this answers is not in the document]
Comment (luncliff) [reply CameraXBasic/app/src/androidTest/java/com/android/example/cameraxbasic/CameraPreviewTest.kt]: Thanks for the point. 

`camera` is moved into the `@Test`

IMO placing `onImageAvailable` to test class's method is clear and simple. But it restricts `reader `and `count` to be member variable.

**Comment by owahltinez** [comment]
Thank you! I think it looks good. Can you please confirm that the test runs successfully?

**Comment by owahltinez** [comment]
Also it looks like you need to rebase with the main branch before we can merge it.

**Comment by luncliff** [comment]
> Also it looks like you need to rebase with the main branch before we can merge it.

Great! I already rebased with main in the 68d8a0f, and with my device the `connectedAndroidTest` runs well.

**Comment by owahltinez** [comment]
I'm still seeing "This branch cannot be rebased due to conflicts"

**Comment by luncliff** [comment]
@owahltinez Sorry. I confused between `rebase` and `merge`. I just pushed after `git rebase main`. Did it work well?

**Comment by owahltinez** [comment]
Much better now, thank you!

**Review by owahltinez — APPROVED** [review]

**Pull request merged**